### PR TITLE
[DRAFT] Manual Reloading and Autoswitch Settings

### PR DESCRIPTION
**Pull request by tunbridgep**
This is designed to address #363.

This is my first major open source contribution, please be gentle.

This currently adds 3 new gameplay options:
image

1. No Auto Weapon Switch. This stops automatically switching to a different weapon when you completely run out of ammo with a given weapon.
2. No Auto Reloading. This prevents the game from automatically reloading your weapon once your current clip is empty.
3. No Backpack Reloading. This prevents the game from automatically reloading weapons that are in your backpack. They should remember their existing ammo count.

Current Status:
1. No Auto Weapon Switch. **Completed**
2. No Auto Reloading. **Completed**.
3. No Backpack Reloading. **Not Started**. I don't know how to realistically approach this. I know the Shotgun etc store their current ammo value, but I don't know how to extend it to every weapon or make it generic. I don't even know if that would be viable to do. I can remove this part of the feature request if it turns out to either be too difficult, or out of scope for the project.

### CODE REVIEW

**Comment by tunbridgep** [comment]
Just a comment on this: Now that we have this, plus the potential FOV change (https://github.com/fgsfdsfgs/perfect_dark/issues/96), plus the menu pausing feature (https://github.com/fgsfdsfgs/perfect_dark/pull/358), it possibly doesn't make sense to put everything in the "Extended/Player %d/Game" options menu, as these settings are global gameplay settings, they aren't dependent on the individual players.

Proposal: What if there was a global "Gameplay Enhancements" menu in the Extended menu, which housed all the big gameplay change stuff - menu pausing, auto reloading, etc - that affects the whole game, so we don't clutter the per-user gameplay settings (crosshair colour etc).

How are these sorts of gameplay modifiers going to affect networked games?

**Comment by fgsfdsfgs** [comment]
Yeah, there should be a "gameplay settings" or "compatibility settings" menu like in Doom source ports. With toggles for stuff like this. They can be synced in netgames in theory to host's settings, though neither pauseless menus nor autoreloading really require any sync (menus are already pauseless in netgames and reloading is signaled when the actual reload happens regardless of what triggered it).

**Comment by tunbridgep** [comment]
Thanks for the response. Would you be up to helping me with the "can't reload when empty" issue? I'm having some trouble understanding the code. Between the strange function names (presumably from the decompiler) and the very complex/convoluted logic, it's very hard to follow what's going on.

I don't even know where to start with the backpack thing.

> though neither pauseless menus nor autoreloading really require any sync (menus are already pauseless in netgames and reloading is signaled when the actual reload happens regardless of what triggered it).

I would argue manual reloading makes the game harder, and could disadvantage a player, therefore it should be a synced server option. No Backpack Reloading DEFINITELY disadvantages a player.

**Comment by 2023PerfectDark** [comment]
You also might want to look into being able to swap weapons while reloading for this too.

**Comment by cylonicboom** [comment]
> This is designed to address #363.
> 
> This is my first major open source contribution, please be gentle.
> 
> This currently adds 3 new gameplay options: image
> 
>     1. No Auto Weapon Switch. This stops automatically switching to a different weapon when you completely run out of ammo with a given weapon.
> 
>     2. No Auto Reloading. This prevents the game from automatically reloading your weapon once your current clip is empty.
> 
>     3. No Backpack Reloading. This prevents the game from automatically reloading weapons that are in your backpack. They should remember their existing ammo count.
> 
> 
> Current Status:
> 
>     1. No Auto Weapon Switch. **Completed**
> 
>     2. No Auto Reloading. **Partially Completed**. Functionality is in, but there's currently an issue where you can't manually reload when your clip is completely empty. I need some help with this.
> 
>     3. No Backpack Reloading. **Not Started**. I don't know how to realistically approach this. I know the Shotgun etc store their current ammo value, but I don't know how to extend it to every weapon or make it generic. I don't even know if that would be viable to do. I can remove this part of the feature request if it turns out to either be too difficult, or out of scope for the project.

nit: I've always heard it referred to as "quick reloading" but this is actually self-descriptive to people not super-versed in pd.

**Comment by tunbridgep** [comment]
Does someone feel like taking this code on? I can't figure out what to do next.

**Comment by tunbridgep** [comment]
Okay, I have figured out the reloading issue. Please review the code and test. I may move the No Backpack Reloading thing to a different PR, since it seems like a much harder task.

I would REALLY like if people could test this, so I can clean it up and un-draft this.

@fgsfdsfgs would you like me to add a new Gameplay Settings menu in Extended as part of this PR, since this is the first thing that's likely going to use it?

**Comment by tunbridgep** [comment]
Sorry, I kind of dropped the ball on this one.

I might finish it up and un-draft it in the next week or so.
